src/algebra/module/graded_module.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import ring_theory.graded_algebra.basic
+import algebra.direct_sum.decomposition
+
+/-!
+# Graded Module
+
+Given a graded `R`-algebra `A` with grading `𝓐 : ι → submodule R A` and an `A`-module `M` with
+decomposition `ι → addsubmonoid M`,
+we define the typeclass `graded_module 𝓐 𝓜` for internally graded modules.
+
+## Main definitions
+
+* `graded_module 𝓐 𝓜`: an `A`-module `M` is graded by `𝓜` if and only if the decomposition map
+`M → ⨁ i, 𝓜 i` is inverse to the canonical map `⨁ i, 𝓜 i → M`.
+* `graded_module.decompose`: `M` and `⨁ i, 𝓜 i` are isomorphic as `add_comm_monoid`.
+* `graded_module.is_module`: `⨁ i, 𝓜 i` is an `A`-module.
+
+## Tags
+
+graded module
+-/
+
+
+open_locale direct_sum big_operators
+
+variables {ι R A : Type*}
+variables [decidable_eq ι] [add_monoid ι] [comm_semiring R] [semiring A] [algebra R A]
+variables (𝓐 : ι → submodule R A) [graded_algebra 𝓐]
+variables {M : Type*} [add_comm_monoid M] [module A M]
+variables (𝓜 : ι → add_submonoid M)
```

Comment:
If possible, this should use `add_submonoid_class` like `decomposition` does.

Reply:
done